File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic.Syntax/TryCastExpressionSyntax.cs
using Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax;

namespace Microsoft.CodeAnalysis.VisualBasic.Syntax
{
	public sealed class TryCastExpressionSyntax : CastExpressionSyntax
	{
		public new SyntaxToken Keyword => new SyntaxToken(this, ((Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.TryCastExpressionSyntax)base.Green)._keyword, base.Position, 0);

		public new SyntaxToken OpenParenToken => new SyntaxToken(this, ((Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.TryCastExpressionSyntax)base.Green)._openParenToken, GetChildPosition(1), GetChildIndex(1));

		public new ExpressionSyntax Expression => GetRed(ref _expression, 2);

		public new SyntaxToken CommaToken => new SyntaxToken(this, ((Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.TryCastExpressionSyntax)base.Green)._commaToken, GetChildPosition(3), GetChildIndex(3));

		public new TypeSyntax Type => GetRed(ref _type, 4);

		public new SyntaxToken CloseParenToken => new SyntaxToken(this, ((Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.TryCastExpressionSyntax)base.Green)._closeParenToken, GetChildPosition(5), GetChildIndex(5));

		internal TryCastExpressionSyntax(GreenNode green, SyntaxNode parent, int startLocation)
			: base(green, parent, startLocation)
		{
		}

		internal TryCastExpressionSyntax(SyntaxKind kind, DiagnosticInfo[] errors, SyntaxAnnotation[] annotations, KeywordSyntax keyword, PunctuationSyntax openParenToken, ExpressionSyntax expression, PunctuationSyntax commaToken, TypeSyntax type, PunctuationSyntax closeParenToken)
			: this(new Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.TryCastExpressionSyntax(kind, errors, annotations, keyword, openParenToken, (Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ExpressionSyntax)expression.Green, commaToken, (Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.TypeSyntax)type.Green, closeParenToken), null, 0)
		{
		}

		internal override SyntaxToken GetKeywordCore()
		{
			return Keyword;
		}

		internal override CastExpressionSyntax WithKeywordCore(SyntaxToken keyword)
		{
			return WithKeyword(keyword);
		}

		public new TryCastExpressionSyntax WithKeyword(SyntaxToken keyword)
		{
			return Update(keyword, OpenParenToken, Expression, CommaToken, Type, CloseParenToken);
		}

		internal override SyntaxToken GetOpenParenTokenCore()
		{
			return OpenParenToken;
		}

		internal override CastExpressionSyntax WithOpenParenTokenCore(SyntaxToken openParenToken)
		{
			return WithOpenParenToken(openParenToken);
		}

		public new TryCastExpressionSyntax WithOpenParenToken(SyntaxToken openParenToken)
		{
			return Update(Keyword, openParenToken, Expression, CommaToken, Type, CloseParenToken);
		}

		internal override ExpressionSyntax GetExpressionCore()
		{
			return Expression;
		}

		internal override CastExpressionSyntax WithExpressionCore(ExpressionSyntax expression)
		{
			return WithExpression(expression);
		}

		public new TryCastExpressionSyntax WithExpression(ExpressionSyntax expression)
		{
			return Update(Keyword, OpenParenToken, expression, CommaToken, Type, CloseParenToken);
		}

		internal override SyntaxToken GetCommaTokenCore()
		{
			return CommaToken;
		}

		internal override CastExpressionSyntax WithCommaTokenCore(SyntaxToken commaToken)
		{
			return WithCommaToken(commaToken);
		}

		public new TryCastExpressionSyntax WithCommaToken(SyntaxToken commaToken)
		{
			return Update(Keyword, OpenParenToken, Expression, commaToken, Type, CloseParenToken);
		}

		internal override TypeSyntax GetTypeCore()
		{
			return Type;
		}

		internal override CastExpressionSyntax WithTypeCore(TypeSyntax type)
		{
			return WithType(type);
		}

		public new TryCastExpressionSyntax WithType(TypeSyntax type)
		{
			return Update(Keyword, OpenParenToken, Expression, CommaToken, type, CloseParenToken);
		}

		internal override SyntaxToken GetCloseParenTokenCore()
		{
			return CloseParenToken;
		}

		internal override CastExpressionSyntax WithCloseParenTokenCore(SyntaxToken closeParenToken)
		{
			return WithCloseParenToken(closeParenToken);
		}

		public new TryCastExpressionSyntax WithCloseParenToken(SyntaxToken closeParenToken)
		{
			return Update(Keyword, OpenParenToken, Expression, CommaToken, Type, closeParenToken);
		}

		internal override SyntaxNode GetCachedSlot(int i)
		{
			return i switch
			{
				2 => _expression, 
				4 => _type, 
				_ => null, 
			};
		}

		internal override SyntaxNode GetNodeSlot(int i)
		{
			return i switch
			{
				2 => Expression, 
				4 => Type, 
				_ => null, 
			};
		}

		public override TResult Accept<TResult>(VisualBasicSyntaxVisitor<TResult> visitor)
		{
			return visitor.VisitTryCastExpression(this);
		}

		public override void Accept(VisualBasicSyntaxVisitor visitor)
		{
			visitor.VisitTryCastExpression(this);
		}

		public TryCastExpressionSyntax Update(SyntaxToken keyword, SyntaxToken openParenToken, ExpressionSyntax expression, SyntaxToken commaToken, TypeSyntax type, SyntaxToken closeParenToken)
		{
			if (keyword != Keyword || openParenToken != OpenParenToken || expression != Expression || commaToken != CommaToken || type != Type || closeParenToken != CloseParenToken)
			{
				TryCastExpressionSyntax tryCastExpressionSyntax = SyntaxFactory.TryCastExpression(keyword, openParenToken, expression, commaToken, type, closeParenToken);
				SyntaxAnnotation[] annotations = GetAnnotations();
				if (annotations != null && annotations.Length > 0)
				{
					return SyntaxNodeExtensions.WithAnnotations(tryCastExpressionSyntax, annotations);
				}
				return tryCastExpressionSyntax;
			}
			return this;
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/Syntax/TryCastExpressionSyntax.vb
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
Imports System

Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax
	Public NotInheritable Class TryCastExpressionSyntax
		Inherits Microsoft.CodeAnalysis.VisualBasic.Syntax.CastExpressionSyntax
		Public Shadows ReadOnly Property CloseParenToken As Microsoft.CodeAnalysis.SyntaxToken
			Get
				Return New Microsoft.CodeAnalysis.SyntaxToken(Me, DirectCast(MyBase.Green, Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.TryCastExpressionSyntax)._closeParenToken, Me.GetChildPosition(5), MyBase.GetChildIndex(5))
			End Get
		End Property

		Public Shadows ReadOnly Property CommaToken As Microsoft.CodeAnalysis.SyntaxToken
			Get
				Return New Microsoft.CodeAnalysis.SyntaxToken(Me, DirectCast(MyBase.Green, Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.TryCastExpressionSyntax)._commaToken, Me.GetChildPosition(3), MyBase.GetChildIndex(3))
			End Get
		End Property

		Public Shadows ReadOnly Property Expression As Microsoft.CodeAnalysis.VisualBasic.Syntax.ExpressionSyntax
			Get
				Return MyBase.GetRed(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.ExpressionSyntax)(Me._expression, 2)
			End Get
		End Property

		Public Shadows ReadOnly Property Keyword As Microsoft.CodeAnalysis.SyntaxToken
			Get
				Return New Microsoft.CodeAnalysis.SyntaxToken(Me, DirectCast(MyBase.Green, Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.TryCastExpressionSyntax)._keyword, MyBase.Position, 0)
			End Get
		End Property

		Public Shadows ReadOnly Property OpenParenToken As Microsoft.CodeAnalysis.SyntaxToken
			Get
				Return New Microsoft.CodeAnalysis.SyntaxToken(Me, DirectCast(MyBase.Green, Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.TryCastExpressionSyntax)._openParenToken, Me.GetChildPosition(1), MyBase.GetChildIndex(1))
			End Get
		End Property

		Public Shadows ReadOnly Property Type As Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeSyntax
			Get
				Return MyBase.GetRed(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeSyntax)(Me._type, 4)
			End Get
		End Property

		Friend Sub New(ByVal green As GreenNode, ByVal parent As SyntaxNode, ByVal startLocation As Integer)
			MyBase.New(green, parent, startLocation)
		End Sub

		Friend Sub New(ByVal kind As SyntaxKind, ByVal errors As DiagnosticInfo(), ByVal annotations As SyntaxAnnotation(), ByVal keyword As KeywordSyntax, ByVal openParenToken As PunctuationSyntax, ByVal expression As Microsoft.CodeAnalysis.VisualBasic.Syntax.ExpressionSyntax, ByVal commaToken As PunctuationSyntax, ByVal type As Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeSyntax, ByVal closeParenToken As PunctuationSyntax)
			MyClass.New(New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.TryCastExpressionSyntax(kind, errors, annotations, keyword, openParenToken, DirectCast(expression.Green, Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ExpressionSyntax), commaToken, DirectCast(type.Green, Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.TypeSyntax), closeParenToken), Nothing, 0)
		End Sub

		Public Overrides Function Accept(Of TResult)(ByVal visitor As VisualBasicSyntaxVisitor(Of TResult)) As TResult
			Return visitor.VisitTryCastExpression(Me)
		End Function

		Public Overrides Sub Accept(ByVal visitor As Microsoft.CodeAnalysis.VisualBasic.VisualBasicSyntaxVisitor)
			visitor.VisitTryCastExpression(Me)
		End Sub

		Friend Overrides Function GetCachedSlot(ByVal i As Integer) As Microsoft.CodeAnalysis.SyntaxNode
			Dim syntaxNode As Microsoft.CodeAnalysis.SyntaxNode
			Dim num As Integer = i
			If (num = 2) Then
				syntaxNode = Me._expression
			ElseIf (num = 4) Then
				syntaxNode = Me._type
			Else
				syntaxNode = Nothing
			End If
			Return syntaxNode
		End Function

		Friend Overrides Function GetCloseParenTokenCore() As Microsoft.CodeAnalysis.SyntaxToken
			Return Me.CloseParenToken
		End Function

		Friend Overrides Function GetCommaTokenCore() As Microsoft.CodeAnalysis.SyntaxToken
			Return Me.CommaToken
		End Function

		Friend Overrides Function GetExpressionCore() As Microsoft.CodeAnalysis.VisualBasic.Syntax.ExpressionSyntax
			Return Me.Expression
		End Function

		Friend Overrides Function GetKeywordCore() As Microsoft.CodeAnalysis.SyntaxToken
			Return Me.Keyword
		End Function

		Friend Overrides Function GetNodeSlot(ByVal i As Integer) As SyntaxNode
			Dim expression As SyntaxNode
			Dim num As Integer = i
			If (num = 2) Then
				expression = Me.Expression
			ElseIf (num = 4) Then
				expression = Me.Type
			Else
				expression = Nothing
			End If
			Return expression
		End Function

		Friend Overrides Function GetOpenParenTokenCore() As Microsoft.CodeAnalysis.SyntaxToken
			Return Me.OpenParenToken
		End Function

		Friend Overrides Function GetTypeCore() As Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeSyntax
			Return Me.Type
		End Function

		Public Function Update(ByVal keyword As Microsoft.CodeAnalysis.SyntaxToken, ByVal openParenToken As Microsoft.CodeAnalysis.SyntaxToken, ByVal expression As Microsoft.CodeAnalysis.VisualBasic.Syntax.ExpressionSyntax, ByVal commaToken As Microsoft.CodeAnalysis.SyntaxToken, ByVal type As Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeSyntax, ByVal closeParenToken As Microsoft.CodeAnalysis.SyntaxToken) As Microsoft.CodeAnalysis.VisualBasic.Syntax.TryCastExpressionSyntax
			Dim tryCastExpressionSyntax As Microsoft.CodeAnalysis.VisualBasic.Syntax.TryCastExpressionSyntax
			If (keyword <> Me.Keyword OrElse openParenToken <> Me.OpenParenToken OrElse expression <> Me.Expression OrElse commaToken <> Me.CommaToken OrElse type <> Me.Type OrElse closeParenToken <> Me.CloseParenToken) Then
				Dim tryCastExpressionSyntax1 As Microsoft.CodeAnalysis.VisualBasic.Syntax.TryCastExpressionSyntax = Microsoft.CodeAnalysis.VisualBasic.SyntaxFactory.TryCastExpression(keyword, openParenToken, expression, commaToken, type, closeParenToken)
				Dim annotations As SyntaxAnnotation() = MyBase.GetAnnotations()
				tryCastExpressionSyntax = If(annotations Is Nothing OrElse CInt(annotations.Length) <= 0, tryCastExpressionSyntax1, tryCastExpressionSyntax1.WithAnnotations(annotations))
			Else
				tryCastExpressionSyntax = Me
			End If
			Return tryCastExpressionSyntax
		End Function

		Public Shadows Function WithCloseParenToken(ByVal closeParenToken As Microsoft.CodeAnalysis.SyntaxToken) As Microsoft.CodeAnalysis.VisualBasic.Syntax.TryCastExpressionSyntax
			Return Me.Update(Me.Keyword, Me.OpenParenToken, Me.Expression, Me.CommaToken, Me.Type, closeParenToken)
		End Function

		Friend Overrides Function WithCloseParenTokenCore(ByVal closeParenToken As Microsoft.CodeAnalysis.SyntaxToken) As Microsoft.CodeAnalysis.VisualBasic.Syntax.CastExpressionSyntax
			Return Me.WithCloseParenToken(closeParenToken)
		End Function

		Public Shadows Function WithCommaToken(ByVal commaToken As Microsoft.CodeAnalysis.SyntaxToken) As Microsoft.CodeAnalysis.VisualBasic.Syntax.TryCastExpressionSyntax
			Return Me.Update(Me.Keyword, Me.OpenParenToken, Me.Expression, commaToken, Me.Type, Me.CloseParenToken)
		End Function

		Friend Overrides Function WithCommaTokenCore(ByVal commaToken As Microsoft.CodeAnalysis.SyntaxToken) As Microsoft.CodeAnalysis.VisualBasic.Syntax.CastExpressionSyntax
			Return Me.WithCommaToken(commaToken)
		End Function

		Public Shadows Function WithExpression(ByVal expression As Microsoft.CodeAnalysis.VisualBasic.Syntax.ExpressionSyntax) As Microsoft.CodeAnalysis.VisualBasic.Syntax.TryCastExpressionSyntax
			Return Me.Update(Me.Keyword, Me.OpenParenToken, expression, Me.CommaToken, Me.Type, Me.CloseParenToken)
		End Function

		Friend Overrides Function WithExpressionCore(ByVal expression As Microsoft.CodeAnalysis.VisualBasic.Syntax.ExpressionSyntax) As Microsoft.CodeAnalysis.VisualBasic.Syntax.CastExpressionSyntax
			Return Me.WithExpression(expression)
		End Function

		Public Shadows Function WithKeyword(ByVal keyword As Microsoft.CodeAnalysis.SyntaxToken) As Microsoft.CodeAnalysis.VisualBasic.Syntax.TryCastExpressionSyntax
			Return Me.Update(keyword, Me.OpenParenToken, Me.Expression, Me.CommaToken, Me.Type, Me.CloseParenToken)
		End Function

		Friend Overrides Function WithKeywordCore(ByVal keyword As Microsoft.CodeAnalysis.SyntaxToken) As Microsoft.CodeAnalysis.VisualBasic.Syntax.CastExpressionSyntax
			Return Me.WithKeyword(keyword)
		End Function

		Public Shadows Function WithOpenParenToken(ByVal openParenToken As Microsoft.CodeAnalysis.SyntaxToken) As Microsoft.CodeAnalysis.VisualBasic.Syntax.TryCastExpressionSyntax
			Return Me.Update(Me.Keyword, openParenToken, Me.Expression, Me.CommaToken, Me.Type, Me.CloseParenToken)
		End Function

		Friend Overrides Function WithOpenParenTokenCore(ByVal openParenToken As Microsoft.CodeAnalysis.SyntaxToken) As Microsoft.CodeAnalysis.VisualBasic.Syntax.CastExpressionSyntax
			Return Me.WithOpenParenToken(openParenToken)
		End Function

		Public Shadows Function WithType(ByVal type As Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeSyntax) As Microsoft.CodeAnalysis.VisualBasic.Syntax.TryCastExpressionSyntax
			Return Me.Update(Me.Keyword, Me.OpenParenToken, Me.Expression, Me.CommaToken, type, Me.CloseParenToken)
		End Function

		Friend Overrides Function WithTypeCore(ByVal type As Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeSyntax) As Microsoft.CodeAnalysis.VisualBasic.Syntax.CastExpressionSyntax
			Return Me.WithType(type)
		End Function
	End Class
End Namespace